cpp/src/io/json/nested_json.hpp

```diff
@@ -183,6 +186,43 @@ struct device_json_column {
   }
 };
 
+namespace experimental {
+struct column_tree_csr {
+  // position of nnzs
+  rmm::device_uvector<NodeIndexT> rowidx;
+  rmm::device_uvector<NodeIndexT> colidx;
+  // node properties
+  rmm::device_uvector<NodeIndexT> column_ids;
+  rmm::device_uvector<NodeT> categories;
+  rmm::device_uvector<SymbolOffsetT> range_begin;
+  rmm::device_uvector<SymbolOffsetT> range_end;
+};
+
+namespace detail {
+/**
+ * @brief Reduce node tree into column tree by aggregating each property of column.
+ *
+ * @param tree json node tree to reduce (modified in-place, but restored to original state)
+ * @param col_ids column ids of each node (modified in-place, but restored to original state)
+ * @param row_offsets row offsets of each node (modified in-place, but restored to original state)
+ * @param stream The CUDA stream to which kernels are dispatched
+ * @return A tuple containing the column tree, identifier for each column and the maximum row index
+ * in each column
+ */
+
+std::tuple<column_tree_csr, rmm::device_uvector<size_type>> reduce_to_column_tree_csr(
+  tree_meta_t& tree,
+  device_span<NodeIndexT> original_col_ids,
+  device_span<NodeIndexT> sorted_col_ids,
+  device_span<NodeIndexT> ordered_node_ids,
+  device_span<size_type> row_offsets,
```

Comment:
Similar to my suggestion above, if we can use array of struct it is better. Each `device_span` cost 16 bytes/2 registers so reducing the number of register usage is very helpful.

Reply:
The number of parameters have been reduced.